go/ql/lib/semmle/go/frameworks/Fasthttp.qll

```diff
@@ -0,0 +1,455 @@
+/**
+ * Provides classes for working with untrusted flow sources, sinks and taint propagators
+ * from the `github.com/valyala/fasthttp` package.
+ */
+
+import go
+private import semmle.go.security.RequestForgeryCustomizations
+
+/**
+ * Provides classes for working with the [fasthttp](github.com/valyala/fasthttp) package.
+ */
+module Fasthttp {
+  /** Gets the v1 module path `github.com/valyala/fasthttp`. */
+  string v1modulePath() { result = "github.com/valyala/fasthttp" }
+
+  /** Gets the path for the root package of fasthttp. */
+  string packagePath() { result = package(v1modulePath(), "") }
+
+  /**
+   * Provide models for sanitizer/Dangerous Functions of fasthttp
+   */
+  module Functions {
+    /**
+     * A function that doesn't sanitize user-provided file paths
+     */
+    class FileSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
+      FileSystemAccess() {
+        exists(Function f |
+          f.hasQualifiedName(packagePath(),
+            [
+              "ServeFile", "ServeFileUncompressed", "ServeFileBytes", "ServeFileBytesUncompressed",
+              "SaveMultipartFile"
+            ]) and
+          this = f.getACall()
+        )
+      }
+
+      override DataFlow::Node getAPathArgument() { result = this.getArgument(1) }
+    }
+
+    /**
+     * A function that can be used as a sanitizer for XSS
+     */
+    class HtmlQuoteSanitizer extends SharedXss::Sanitizer {
+      HtmlQuoteSanitizer() {
+        exists(DataFlow::CallNode c |
+          c.getTarget()
+              .hasQualifiedName(packagePath(),
+                ["AppendHTMLEscape", "AppendHTMLEscapeBytes", "AppendQuotedArg"])
+        |
+          this = c.getArgument(1)
+        )
+      }
+    }
+
+    /**
+     * A function that sends HTTP requests.
+     *
+     * Get* send a HTTP GET request.
+     * Post send a HTTP POST request.
+     * These functions first argument is a URL.
+     */
+    class RequestForgerySink extends RequestForgery::Sink {
+      RequestForgerySink() {
+        exists(Function f |
+          f.hasQualifiedName(packagePath(), ["Get", "GetDeadline", "GetTimeout", "Post"]) and
+          this = f.getACall().getArgument(1)
+        )
+      }
+
+      override DataFlow::Node getARequest() { result = this }
+
+      override string getKind() { result = "URL" }
+    }
+
+    /**
+     * A function that sends HTTP requests.
+     * First argument of following functions need Additional steps.
+     * look at URI module, additional steps part for more information.
```

Comment:
If you think we are likely to reference one set of functions and not the other then it is better to not merge the two classes.

I do not know what to do about `getKind`. It is used in the query message in `go/ql/src/Security/CWE-918/RequestForgery.ql`, as "the X of this request". I will think about it.

Reply:
I discussed this with my colleagues. The conclusion was that SSRF is specifically about controlling the destination of the request (and optionally its contents for increased impact), so it’s a requirement that the request URI, and particularly the host part, is under the attacker’s control to raise an alert. Currently the sink is a whole Request object, so we should either add a mandatory taint step (via flow states) through `Request.setRequestURI`, or simply make that the sink. That automatically solves the `getKind` problem, because in SSRF it should always be URL (or host/IP).